app/master/build_scheduler.py

```diff
@@ -112,16 +114,17 @@ def execute_next_subjob_or_free_executor(self, slave):
                 subjob = self._build._unstarted_subjobs.get(block=False)
                 self._logger.debug('Sending {} to {}.', subjob, slave)
                 try:
-                    slave.start_subjob(subjob)
-                    subjob.mark_in_progress(slave)
+                    if self._build.is_canceled:
+                        self._free_slave_executor(slave)
```

Comment:
Could we move this check up, even before acquiring the lock? Maybe like:
```python
if self._build.is_canceled:
    self._free_slave_executor(slave)
    return

try:
    with self._subjob_assignment_lock:
        ...
```
It doesn't really matter, but I feel like we should not even try to get the lock or pull from the unstarted_subjobs queue if the build is already canceled.

I know you don't love multiple return points in a method. 😬 An alternative would be just an if/else around the whole thing.

While we're here, I'm noticing how unnecessarily wide the outer try/except is. The `except Empty` really only applies to that one line `self._build._unstarted_subjobs.get(block=False)`. Not sure if narrowing that try/except would help clean up the flow or not. :woman_shrugging: